model_analyzer/result/concurrency_search.py

```diff
@@ -0,0 +1,155 @@
+# limitations under the License.
+
+from typing import List, Tuple, Optional, Generator
+
+from model_analyzer.model_analyzer_exceptions import TritonModelAnalyzerException
+from model_analyzer.config.input.config_command_profile import ConfigCommandProfile
+from model_analyzer.result.run_config_measurement import RunConfigMeasurement
+
+from math import log2
+
+import logging
+from model_analyzer.constants import LOGGER_NAME, THROUGHPUT_MINIMUM_GAIN, THROUGHPUT_MINIMUM_CONSECUTIVE_CONCURRENCY_TRIES
+
+logger = logging.getLogger(LOGGER_NAME)
+
+
+class ConcurrencySearch():
+    """
+    Determines the next concurrency value to use when searching through
+    RunConfigMeasurements for the best value (according to the users objective)
+      - Will sweep from by powers of two from min to max concurrency
+      - If the user specifies a constraint, the algorithm will perform a binary search 
+        around the boundary if the constraint is violated
+        
+    Invariant: It is necessary for the user to add new measurements as they are taken
+    """
+
+    def __init__(self, config: ConfigCommandProfile) -> None:
+        """
+        Parameters
+        ----------
+        config: ConfigCommandProfile
+            Profile configuration information
+        """
+        self._min_concurrency_index = int(
+            log2(config.run_config_search_min_concurrency))
+        self._max_concurrency_index = int(
+            log2(config.run_config_search_max_concurrency))
+
+        self._run_config_measurements: List[RunConfigMeasurement] = []
+        self._concurrencies: List[int] = []
+        self._last_failing_concurrency = 0
+        self._last_passing_concurrency = 0
+
+    def add_run_config_measurement(
+            self, run_config_measurement: RunConfigMeasurement) -> None:
+        """
+        Adds a new RunConfigMeasurement
+        Invariant: Assumed that RCMs are added in the same order they are measured
+        """
+        self._run_config_measurements.append(run_config_measurement)
+
+    def search_concurrencies(self) -> Generator[int, None, None]:
+        """
+        First performs a concurrency sweep, and then, if necessary, perform
+        a binary concurrency search around the point where the constraint
+        violated
+        """
+        yield from self._perform_concurrency_sweep()
+
+        if self._was_constraint_violated():
+            yield from self._perform_binary_concurrency_search()
+
+    def _perform_concurrency_sweep(self) -> Generator[int, None, None]:
+        for concurrency in (2**i for i in range(
+                self._min_concurrency_index, self._max_concurrency_index + 1)):
+            if self._continue_concurrency_sweep():
+                self._concurrencies.append(concurrency)
+                yield concurrency
+            else:
+                logger.info(
+                    "Terminating concurrency sweep - throughput is decreasing")
+
+    def _continue_concurrency_sweep(self) -> bool:
```

Comment:
This is something I often do. It's because in my mind I've added a `?` to the end of the method name because it gets used in an if-statement.  I've also changed this for `_minimum_tries_reached`